README
- Supported Games

  This table displays which of the main actions are
supported for each game.

Action   1 2 3 4 5 6 7 75 8 9 95 10 105 11 12 123 125 128 13 14 143
thanm x  - - - - - - y  - y y  y  y   -  y  y   -   y   y  y  y   y
thanm l  - - - - - y y  - y y  y  y   -  y  y   -   y   y  y  y   y
thdat x  y y y y y y y  n y y  y  y   n  y  y   n   y   y  y  y   y
thdat c  y y y y y y y  n y y  y  y   n  y  y   n   y   y  y  y   y
thecl d  - - - - - y y  - y y  y  y   -  y  y   -   y   y  y  y   y
thecl c  - - - - - y y  - y y  y  y   -  y  y   -   y   y  y  y   y
thmsg d  - - - - - y y  - y y  y  y   -  y  y   -   y   y  y  y   y
thmsg c  - - - - - y y  - y y  y  y   -  y  y   -   y   y  y  y   y


NEWS
What's new in thtkx 6:

thanm: Initial support for TH14 and TH143 added (some minor issues remains)
thdat: Full support for TH14 and TH143 added
thecl: Full support for TH14 and TH143 added
thmsg: Full support for TH14 and TH143 added


expr.c
/*
 * Redistribution and use in source and binary forms, with
 * or without modification, are permitted provided that the
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain this list
 *    of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce this
 *    list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
#include <config.h>
#include <stddef.h>
#include "ecsparse.h"
#include "expr.h"

typedef struct {
    int symbol;
    int symbols[2];
} alternative_t;

static const alternative_t
th10_alternatives[] = {
    { ADD,      { ADDI, ADDF } },
    { SUBTRACT, { SUBTRACTI, SUBTRACTF } },
    { MULTIPLY, { MULTIPLYI, MULTIPLYF } },
    { DIVIDE,   { DIVIDEI, DIVIDEF } },
    { EQUAL,    { EQUALI, EQUALF } },
    { INEQUAL,  { INEQUALI, INEQUALF } },
    { LT,       { LTI, LTF } },
    { LTEQ,     { LTEQI, LTEQF } },
    { GT,       { GTI, GTF } },
    { GTEQ,     { GTEQI, GTEQF } },
    { 0,        { 0, 0 } }
};

static const expr_t
th10_expressions[] = {
    /* The program checks against the number of params, as well as the
     * requested stack depth, and does the replacements. */
    /* p0 is the first param, p1 the second ... */
    /* s0 is the previous instruction, s1 the one previous to s0 ... */

    /*SYM         ID  RET     P  A    S   DISP */
    { GOTO,       12,   0, "oS", 0,  "S",           "goto p0 @ p1" },
    { UNLESS,     13,   0, "oS", 1,  "S", "unless s0 goto p0 @ p1" },
    { IF,         14,   0, "oS", 1,  "S",     "if s0 goto p0 @ p1" },

    { LOADI,      42, 'S',  "S", 0, NULL, "p0" },
    { ASSIGNI,    43,   0,  "S", 1,  "S", "p0 = s0" },
    { LOADF,      44, 'f',  "f", 0, NULL, "p0" },
    { ASSIGNF,    45,   0,  "f", 1,  "f", "p0 = s0" },

    { ADDI,       50, 'S', NULL, 2, "SS", "s1 + s0" },
    { ADDF,       51, 'f', NULL, 2, "ff", "s1 + s0" },
    { SUBTRACTI,  52, 'S', NULL, 2, "SS", "s1 - s0" },
    { SUBTRACTF,  53, 'f', NULL, 2, "ff", "s1 - s0" },
    { MULTIPLYI,  54, 'S', NULL, 2, "SS", "s1 * s0" },
    { MULTIPLYF,  55, 'f', NULL, 2, "ff", "s1 * s0" },
    { DIVIDEI,    56, 'S', NULL, 2, "SS", "s1 / s0" },
    { DIVIDEF,    57, 'f', NULL, 2, "ff", "s1 / s0" },
    { MODULO,     58, 'S', NULL, 2, "SS", "s1 % s0" },
    { EQUALI,     59, 'S', NULL, 2, "SS", "s1 == s0" },
    { EQUALF,     60, 'S', NULL, 2, "ff", "s1 == s0" },
    { INEQUALI,   61, 'S', NULL, 2, "SS", "s1 != s0" },
    { INEQUALF,   62, 'S', NULL, 2, "ff", "s1 != s0" },
    { LTI,        63, 'S', NULL, 2, "SS", "s1 < s0" },
    { LTF,        64, 'S', NULL, 2, "ff", "s1 < s0" },
    { LTEQI,      65, 'S', NULL, 2, "SS", "s1 <= s0" },
    { LTEQF,      66, 'S', NULL, 2, "ff", "s1 <= s0" },
    { GTI,        67, 'S', NULL, 2, "SS", "s1 > s0" },
    { GTF,        68, 'S', NULL, 2, "ff", "s1 > s0" },
    { GTEQI,      69, 'S', NULL, 2, "SS", "s1 >= s0" },
    { GTEQF,      70, 'S', NULL, 2, "ff", "s1 >= s0" },
    { NOT,        71, 'S', NULL, 1, "SS", "!s0" },
/*  { XXX,        72,   0, NULL, 0, NULL, NULL },*/
    { OR,         73, 'S', NULL, 2, "SS", "s1 || s0" },
    { AND,        74, 'S', NULL, 2, "SS", "s1 && s0" },
    { XOR,        75, 'S', NULL, 2, "SS", "s1 ^ s0" },
    { 0,           0,   0, NULL, 0, NULL, NULL }
};

static const expr_t*
expr_get_table(unsigned int version)
{
    if (   version != 10
        && version != 11
        && version != 12
        && version != 125
        && version != 128
        && version != 13
        && version != 14
        && version != 143)
        return NULL;
    return th10_expressions;
}

const expr_t*
expr_get_by_symbol(
    unsigned int version,
    int symbol)
{
    const expr_t* table = expr_get_table(version);

    while (table->symbol) {
        if (table->symbol == symbol)
            return table;
        ++table;
    }

    return NULL;
}

const expr_t*
expr_get_by_id(
    unsigned int version,
    int id)
{
    const expr_t* table = expr_get_table(version);

    while (table->symbol) {
        if (table->id == id)
            return table;
        ++table;
    }

    return NULL;
}

int
expr_is_leaf(
    unsigned int version,
    int id)
{
    const expr_t* expr = expr_get_by_id(version, id);

    if (!expr)
        return 0;

    return expr->stack_arity == 0 &&
           expr->return_type != 0;
}


thanm.h
/*
 * Redistribution and use in source and binary forms, with
 * or without modification, are permitted provided that the
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain this list
 *    of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce this
 *    list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
#ifndef THANM_H_
#define THANM_H_

#include <config.h>
#include <inttypes.h>
#include "list.h"

extern unsigned int option_force;

typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    uint32_t id;
    float x, y, w, h;
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
} PACK_ATTRIBUTE sprite_t;

typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    uint16_t time;
    uint8_t type;
    /* XXX: data length. */
    uint8_t length;
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
    unsigned char data[];
} PACK_ATTRIBUTE anm_instr0_t;

typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    uint16_t type;
    uint16_t length;
    uint16_t time;
    /* TODO: Implement this, it works similarly to that one in ECL files. */
    uint16_t param_mask;
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
    unsigned char data[];
} PACK_ATTRIBUTE anm_instr_t;

/* TODO: Rename this struct to _header_t something. */
typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    int32_t id;
    uint32_t offset;
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
} PACK_ATTRIBUTE anm_offset_t;

typedef struct {
    anm_offset_t* offset;
    list_t instrs;
} anm_script_t;

typedef struct {
    uint16_t type;
    const char* format;
} opcode_fmt_t;

typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    uint32_t sprites;
    uint32_t scripts;
    uint32_t zero1;
    uint32_t w, h;
    uint32_t format;
    uint32_t zero2; /* XXX: A few are 0xff000000. */
    uint32_t nameoffset;
    /* XXX: X is unused here. */
    /* XXX: Y stores the secondary name offset for TH06.
     *      There is no secondary name when it is zero. */
    uint32_t x, y;
    /* 0: TH06
     * 2: TH07
     * 3: TH08, TH09
     * 4: TH95, TH10
     * 7: TH11, TH12, TH125 */
    uint32_t version;
    /* 0  - Random things, everything in TH06. */
    /* 1  - data/ascii/loading.png for TH08 and TH09. */
    /* 10 - Mostly sprites? */
    /* 11 - Used mainly for backgrounds and ascii.png. */
    uint32_t unknown1;
    uint32_t thtxoffset;
    uint16_t hasdata;
    uint16_t unknown2;
    uint32_t nextoffset;
    uint32_t zero3;
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
} PACK_ATTRIBUTE anm_header_t;

typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    uint32_t version;
    uint16_t sprites;
    uint16_t scripts;
    uint16_t zero1;
    uint16_t w, h;
    uint16_t format;
    uint32_t nameoffset;
    uint16_t x, y;
    uint32_t unknown1;
    uint32_t thtxoffset;
    uint16_t hasdata;
    uint16_t unknown2;
    uint32_t nextoffset;
    uint32_t zero2[6];
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
} PACK_ATTRIBUTE anm_header11_t;

typedef struct {
#ifdef PACK_PRAGMA
#pragma pack(push,1)
#endif
    char magic[4];
    uint16_t zero;
    uint16_t format;
    /* These may be different from the parent entry. */
    uint16_t w, h;
    uint32_t size;
#ifdef PACK_PRAGMA
#pragma pack(pop)
#endif
    unsigned char data[];
} PACK_ATTRIBUTE thtx_header_t;

typedef struct {
    anm_header_t* header;
    thtx_header_t* thtx;

    char* name;
    char* name2;

    /* List of sprite_t*. */
    list_t sprites;
    /* List of anm_script_t*. */
    list_t scripts;

    unsigned char* data;
} anm_entry_t;

typedef struct {
    unsigned char* map;
    long map_size;

    /* List of const char*. */
    list_t names;
    /* List of anm_entry_t*. */
    list_t entries;
} anm_archive_t;

#endif


thecl.c
/*
 * Redistribution and use in source and binary forms, with
 * or without modification, are permitted provided that the
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain this list
 *    of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce this
 *    list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
#include <config.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "program.h"
#include "thecl.h"
#include "util.h"

extern const thecl_module_t th06_ecl;
extern const thecl_module_t th10_ecl;

thecl_t*
thecl_new(
    void)
{
    thecl_t* ecl = calloc(1, sizeof(thecl_t));
    list_init(&ecl->subs);
    list_init(&ecl->local_data);
    return ecl;
}

static void
thecl_free(
    thecl_t* ecl)
{
    for (size_t a = 0; a < ecl->anim_count; ++a)
        free(ecl->anim_names[a]);
    free(ecl->anim_names);

    for (size_t e = 0; e < ecl->ecli_count; ++e)
        free(ecl->ecli_names[e]);
    free(ecl->ecli_names);

    thecl_sub_t* sub;
    list_for_each(&ecl->subs, sub) {
        free(sub->name);

        thecl_instr_t* instr;
        list_for_each(&sub->instrs, instr)
            thecl_instr_free(instr);
        list_free_nodes(&sub->instrs);

        for (size_t v = 0; v < sub->var_count; ++v)
            free(sub->vars[v]);
        free(sub->vars);

        thecl_label_t* label;
        list_for_each(&sub->labels, label)
            free(label);
        list_free_nodes(&sub->labels);

        free(sub);
    }
    list_free_nodes(&ecl->subs);

    thecl_local_data_t* local_data;
    list_for_each(&ecl->local_data, local_data)
        free(local_data);
    list_free_nodes(&ecl->local_data);

    free(ecl);
}

thecl_instr_t*
thecl_instr_new(void)
{
    thecl_instr_t* instr = calloc(1, sizeof(thecl_instr_t));
    instr->type = THECL_INSTR_INSTR;
    list_init(&instr->params);
    return instr;
}

thecl_instr_t*
thecl_instr_time(unsigned int time)
{
    thecl_instr_t* instr = thecl_instr_new();
    instr->type = THECL_INSTR_TIME;
    instr->time = time;
    return instr;
}

thecl_instr_t*
thecl_instr_rank(unsigned int rank)
{
    thecl_instr_t* instr = thecl_instr_new();
    instr->type = THECL_INSTR_RANK;
    instr->rank = rank;
    return instr;
}

thecl_instr_t*
thecl_instr_label(unsigned int offset)
{
    thecl_instr_t* instr = thecl_instr_new();
    instr->type = THECL_INSTR_LABEL;
    instr->offset = offset;
    return instr;
}

void
thecl_instr_free(thecl_instr_t* instr)
{
    free(instr->string);

    thecl_param_t* param;
    list_for_each(&instr->params, param) {
        value_free(&param->value);
        free(param);
    }
    list_free_nodes(&instr->params);

    free(instr);
}

thecl_param_t*
param_new(
    int type)
{
    thecl_param_t* param = calloc(1, sizeof(thecl_param_t));
    param->type = type;
    param->value.type = type;
    return param;
}

void
param_free(
    thecl_param_t* param)
{
    value_free(&param->value);
    free(param);
}

static void
print_usage(void)
{
    printf("Usage: %s COMMAND[OPTION...] [INPUT [OUTPUT]]\n"
           "COMMAND can be:\n"
           "  c  create ECL file\n"
           "  d  dump ECL file\n"
           "  V  display version information and exit\n"
           "OPTION can be:\n"
           "  #  # can be 6, 7, 8, 9, 95, 10, 11, 12, 125, 128, 13, 14, or 143 (required)\n"
           "Report bugs to <" PACKAGE_BUGREPORT ">.\n", argv0);
}

int
main(int argc, char* argv[])
{
    FILE* in = stdin;
    FILE* out = stdout;
    unsigned int version = 0;
    int mode;
    const thecl_module_t* module = NULL;

    current_input = "(stdin)";
    current_output = "(stdout)";

    mode = parse_args(argc, argv, print_usage, "cdV", "", &version);

    if (!mode)
        exit(1);

    switch (version) {
    case 6:
    case 7:
    case 8:
    case 9:
    case 95:
        module = &th06_ecl;
        break;
    case 10:
    case 11:
    case 12:
    case 125:
    case 128:
    case 13:
    case 14:
    case 143:
        module = &th10_ecl;
        break;
    default:
        if (mode == 'c' || mode == 'd') {
            if (version == 0)
                fprintf(stderr, "%s: version must be specified\n", argv0);
            else
                fprintf(stderr, "%s: version %u is unsupported\n", argv0, version);
            exit(1);
        }
    }

    switch (mode)
    {
    case 'c':
    case 'd': {
        if (argc > 2) {
            current_input = argv[2];
            in = fopen(argv[2], "rb");
            if (!in) {
                fprintf(stderr, "%s: couldn't open %s for reading: %s\n",
                    argv0, argv[2], strerror(errno));
                exit(1);
            }
            if (argc > 3) {
                current_output = argv[3];
                out = fopen(argv[3], "wb");
                if (!out) {
                    fprintf(stderr, "%s: couldn't open %s for writing: %s\n",
                        argv0, argv[3], strerror(errno));
                    fclose(in);
                    exit(1);
                }
            }
        }

        if (mode == 'c') {
            thecl_t* ecl = module->parse(in, version);
            if (!ecl)
                exit(1);
            module->compile(ecl, out);
            thecl_free(ecl);
        } else if (mode == 'd') {
            thecl_t* ecl = module->open(in, version);
            if (!ecl)
                exit(1);
            module->trans(ecl);
            module->dump(ecl, out);
            thecl_free(ecl);
        }
        fclose(in);
        fclose(out);

        exit(0);
    }
    case 'V':
        util_print_version();
        exit(0);
    default:
        exit(1);
    }
}


thtk/thdat.c
/*
 * Redistribution and use in source and binary forms, with
 * or without modification, are permitted provided that the
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain this list
 *    of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce this
 *    list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
#include <config.h>
#include <ctype.h>
#ifdef HAVE_LIBGEN_H
#include <libgen.h>
#endif
#include <stdlib.h>
#include <string.h>
#include <thtk/thtk.h>
#include "thdat.h"
#include "thrle.h"

extern const thdat_module_t archive_th02;
extern const thdat_module_t archive_th06;
extern const thdat_module_t archive_th08;
extern const thdat_module_t archive_th95;

static const thdat_module_t*
thdat_version_to_module(
    unsigned int version,
    thtk_error_t** error)
{
    switch (version) {
    case 2:
    case 3:
    case 4:
    case 5:
        return &archive_th02;
    case 6:
    case 7:
        return &archive_th06;
    case 8:
    case 9:
        return &archive_th08;
    case 95:
    case 10:
    case 11:
    case 12:
    case 125:
    case 128:
    case 13:
    case 14:
    case 143:
        return &archive_th95;
    }

    thtk_error_new(error, "no module found for version %u", version);
    return NULL;
}

void
thdat_entry_init(
    thdat_entry_t* entry)
{
    if (entry) {
        memset(entry->name, 0, sizeof(entry->name));
        entry->extra = 0;
        entry->offset = entry->zsize = entry->size = -1;
    }
}

static thdat_t*
thdat_new(
    unsigned int version,
    thtk_io_t* stream,
    thtk_error_t** error)
{
    thdat_t* thdat;
    if (!stream) {
        thtk_error_new(error, "invalid parameter passed");
        return 0;
    }
    thdat = malloc(sizeof(*thdat));
    thdat->version = version;
    if (!(thdat->module = thdat_version_to_module(version, error))) {
        free(thdat);
        return NULL;
    }
    thdat->stream = stream;
    thdat->entry_count = 0;
    thdat->entries = NULL;
    thdat->offset = 0;
    return thdat;
}

thdat_t*
thdat_open(
    unsigned int version,
    thtk_io_t* input,
    thtk_error_t** error)
{
    thdat_t* thdat;
    if (!input) {
        thtk_error_new(error, "invalid parameter passed");
        return 0;
    }
    if (thtk_io_seek(input, 0, SEEK_SET, error) == -1)
        return NULL;
    if (!(thdat = thdat_new(version, input, error)))
        return NULL;
    if (!thdat->module->open(thdat, error)) {
        thdat_free(thdat);
        return NULL;
    }
    return thdat;
}

thdat_t*
thdat_create(
    unsigned int version,
    thtk_io_t* output,
    size_t entry_count,
    thtk_error_t** error)
{
    thdat_t* thdat;
    if (!output) {
        thtk_error_new(error, "invalid parameter passed");
        return 0;
    }
    if (thtk_io_seek(output, 0, SEEK_SET, error) == -1)
        return NULL;
    if (!(thdat = thdat_new(version, output, error)))
        return NULL;
    thdat->entry_count = entry_count;
    thdat->entries = calloc(entry_count, sizeof(thdat_entry_t));
    if (!thdat->module->create(thdat, error)) {
        thdat_free(thdat);
        return NULL;
    }
    return thdat;
}

static int
thdat_entry_compar(
    const void* a,
    const void* b)
{
    const thdat_entry_t* ea = a;
    const thdat_entry_t* eb = b;
    return (int)ea->offset - eb->offset;
}

int
thdat_close(
    thdat_t* thdat,
    thtk_error_t** error)
{
    if (!thdat) {
        thtk_error_new(error, "invalid parameter passed");
        return 0;
    }
    qsort(thdat->entries, thdat->entry_count, sizeof(thdat_entry_t), thdat_entry_compar);
    return thdat->module->close(thdat, error);
}

void
thdat_free(
    thdat_t* thdat)
{
    if (thdat) {
        free(thdat->entries);
        free(thdat);
    }
}

ssize_t
thdat_entry_count(
    thdat_t* thdat,
    thtk_error_t** error)
{
    if (!thdat) {
        thtk_error_new(error, "invalid parameter passed");
        return -1;
    }
    return thdat->entry_count;
}

ssize_t
thdat_entry_by_name(
    thdat_t* thdat,
    const char* name,
    thtk_error_t** error)
{
    if (!thdat || !name) {
        thtk_error_new(error, "invalid parameter passed");
        return -1;
    }
    for (size_t e = 0; e < thdat->entry_count; ++e) {
        if (strcmp(name, thdat->entries[e].name) == 0)
            return e;
    }
    return -1;
}

int
thdat_entry_set_name(
    thdat_t* thdat,
    int entry_index,
    const char* name,
    thtk_error_t** error)
{
    if (thdat && name && entry_index >= 0 && entry_index < (int)thdat->entry_count) {
        char temp_name[256];
        strncpy(temp_name, name, 255);

        if (thdat->module->flags & THDAT_BASENAME) {
#ifdef WIN32
            char filename[_MAX_FNAME];
            char ext[_MAX_EXT];
            _splitpath(temp_name, NULL, NULL, filename, ext);
            snprintf(temp_name, 255, "%s%s", filename, ext);
#else
            char temp_name2[256];
            strncpy(temp_name2, temp_name, 255);
            strncpy(temp_name, basename(temp_name2), 255);
#endif
        }

        if (thdat->module->flags & THDAT_UPPERCASE) {
            for (unsigned int i = 0; i < 255 && temp_name[i]; ++i) {
                temp_name[i] = toupper(temp_name[i]);
            }
        }

        if (thdat->module->flags & THDAT_8_3) {
            const char* dotpos = strchr(temp_name, '.');
            size_t name_len = dotpos ? strlen(temp_name) - strlen(dotpos) : strlen(temp_name);
            size_t ext_len = dotpos ? strlen(dotpos + 1) : 0;

            if (name_len > 8 || ext_len > 3) {
                thtk_error_new(error, "name is not 8.3");
                return 0;
            }
        }

        strcpy(thdat->entries[entry_index].name, temp_name);

        return 1;
    }

    thtk_error_new(error, "invalid parameter passed");
    return 0;
}

const char*
thdat_entry_get_name(
    thdat_t* thdat,
    int entry_index,
    thtk_error_t** error)
{
    if (!thdat || entry_index < 0 || entry_index >= (int)thdat->entry_count) {
        thtk_error_new(error, "invalid parameter passed");
        return NULL;
    }
    return thdat->entries[entry_index].name;
}

ssize_t
thdat_entry_get_size(
    thdat_t* thdat,
    int entry_index,
    thtk_error_t** error)
{
    if (!thdat || entry_index < 0 || entry_index >= (int)thdat->entry_count) {
        thtk_error_new(error, "invalid parameter passed");
        return -1;
    }
    return thdat->entries[entry_index].size;
}

ssize_t
thdat_entry_get_zsize(
    thdat_t* thdat,
    int entry_index,
    thtk_error_t** error)
{
    if (!thdat || entry_index < 0 || entry_index >= (int)thdat->entry_count) {
        thtk_error_new(error, "invalid parameter passed");
        return -1;
    }
    return thdat->entries[entry_index].zsize;
}

ssize_t
thdat_entry_write_data(
    thdat_t* thdat,
    int entry_index,
    thtk_io_t* input,
    size_t input_length,
    thtk_error_t** error)
{
    if (!thdat || entry_index < 0 || entry_index >= (int)thdat->entry_count || !input) {
        thtk_error_new(error, "invalid parameter passed");
        return -1;
    }
    return thdat->module->write(thdat, entry_index, input, input_length, error);
}

ssize_t
thdat_entry_read_data(
    thdat_t* thdat,
    int entry_index,
    thtk_io_t* output,
    thtk_error_t** error)
{
    if (!thdat || entry_index < 0 || entry_index >= (int)thdat->entry_count || !output) {
        thtk_error_new(error, "invalid parameter passed");
        return -1;
    }
    return thdat->module->read(thdat, entry_index, output, error);
}
